src/packet.rs

```diff
@@ -458,16 +387,21 @@ packets! {
             0x3d => DisplayScoreboard { position: i8, score_name: String }
             // 0x3e => UpdateTeam { team_name: String, action: TeamAction }
             0x3f => PluginMessage { channel: String, data: Vec<u8>;
-                impl Packet for PluginMessage {
-                    fn encode(&self, mut dst: &mut Write) -> io::Result<()> {
-                        try!(<String as Protocol>::proto_encode(&self.channel, dst));
-                        try!(dst.write_all(&self.data));
+                impl Protocol for PluginMessage {
+                    type Clean = Self;
+                    fn proto_len(this: &Self) -> usize {
+                        <String as Protocol>::proto_len(&this.channel) + this.data.len()
+                    }
+                    fn proto_encode(this: &Self, mut dst: &mut Write) -> io::Result<()> {
+                        try!(<Var<i32> as Protocol>::proto_encode(&0x3f, dst));
+                        try!(<String as Protocol>::proto_encode(&this.channel, dst));
+                        try!(dst.write_all(&this.data));
                         Ok(())
                     }
-                    fn decode(mut src: &mut Read, len: usize) -> io::Result<PluginMessage> {
+                    fn proto_decode(mut src: &mut Read) -> io::Result<PluginMessage> {
                         Ok(PluginMessage{
                             channel: try!(<String as Protocol>::proto_decode(src)),
-                            data: try!(src.read_exact(len)),
+                            data:  { let mut data = vec![]; try!(src.read_to_end(&mut data)); data },
```

Comment:
This, right now, is not safe, we need `src` to be an adaptor or buffer or bad things could happen. Easy stuff.